### PR TITLE
stop if a keyword is repeated, closes #83

### LINKED ISSUE #83: Multiple definitions of keywords
* parselglossy version: 0.3.0

### Description

Currently the following is allowed
```
Foo {
  bar = true
}

Foo {
  bar = false
}
```
and the last definition takes precedence. This should probably not be allowed.

### PR DESCRIPTION
**Pull request by bast**
if a keyword at the same level is repeated,
the code raises ParselglossyError
with "A keyword is repeaded. Please check your input."

the code does not know which keyword is repeated

### CODE REVIEW

File: parselglossy/grammars/lexer.py [hunk 1]
```diff
@@ -39,6 +39,52 @@
 from . import getkw
 
 
+def flatten_list(nested_list):
```
Comment (robertodr) [inline parselglossy/grammars/lexer.py]: Can you type hints?

File: parselglossy/grammars/lexer.py [hunk 2]
```diff
@@ -39,6 +39,52 @@
...
+    """Flattens a nested list into a flat list.
+
+    Parameters
+    ----------
+    nested_list : Nested list
+
+    Returns
+    -------
+    Flattened list
+    """
```
Comment (robertodr) [inline parselglossy/grammars/lexer.py]: Also in the docs, please adapt to follow the NumPy convention:
```
    Parameters
    ----------
    nested_list : List[Any]
          Nested list

    Returns
    -------
    list: List[Any]
         Flattened list
```

File: parselglossy/grammars/lexer.py [hunk 3]
```diff
@@ -39,6 +39,52 @@
...
+    if nested_list == []:
+        return nested_list
+    if isinstance(nested_list[0], list):
+        return flatten_list(nested_list[0]) + flatten_list(nested_list[1:])
+    return nested_list[:1] + flatten_list(nested_list[1:])
+
+
+def dict_to_list(d):
```
Comment (robertodr) [inline parselglossy/grammars/lexer.py]: Same here, type hints + documentation format.

File: parselglossy/grammars/lexer.py [hunk 4]
```diff
@@ -39,6 +39,52 @@
...
+        else:
+            nested_list.append([k, v])
+    return nested_list
+
+
+def parse_string_to_dict(lexer, s):
+    tokens_dict = lexer.parseString(s).asDict()
+    tokens_list = lexer.parseString(s).asList()
+    if flatten_list(tokens_list) != flatten_list(dict_to_list(tokens_dict)):
+        raise ParselglossyError("A keyword is repeaded. Please check your input.")
```
Comment (robertodr) [inline parselglossy/grammars/lexer.py]: Typo: "repeaded" --> "repeated"

Comment (robertodr) [reply parselglossy/grammars/lexer.py]: This needs to be changed in the tests too.

File: parselglossy/grammars/lexer.py [hunk 5]
```diff
@@ -39,6 +39,52 @@
...
+    nested_list = []
+    for k, v in d.items():
+        if isinstance(v, dict):
+            nested_list.append([k, dict_to_list(v)])
+        else:
+            nested_list.append([k, v])
+    return nested_list
+
+
+def parse_string_to_dict(lexer, s):
```
Comment (robertodr) [inline parselglossy/grammars/lexer.py]: Type hints + documentation.

**Review by robertodr — CHANGES_REQUESTED** [review]
Looks good, we can iterate from here to get a more useful error message. Few changes to make (type hints and docs) plus remember the DCO :)

**Review by robertodr — APPROVED** [review]

**Comment by bast** [comment]
Seems it needs rebase. Weird. I will rebase.

**Comment by robertodr** [comment]
It seems macOS + Python 3.8 is slightly flaky, but it's now all green. Thank you!

**Pull request merged**